pkg/pushsync/pushsync_test.go

```diff
@@ -2,4 +2,336 @@
+
+import (
+	"bytes"
+	"context"
+	"io/ioutil"
+	"runtime"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/pushsync"
+	"github.com/ethersphere/bee/pkg/pushsync/pb"
+
+	"github.com/ethersphere/bee/pkg/addressbook"
+	"github.com/ethersphere/bee/pkg/discovery/mock"
+	"github.com/ethersphere/bee/pkg/localstore"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/p2p"
+	p2pmock "github.com/ethersphere/bee/pkg/p2p/mock"
+	"github.com/ethersphere/bee/pkg/p2p/protobuf"
+	"github.com/ethersphere/bee/pkg/p2p/streamtest"
+	mockstate "github.com/ethersphere/bee/pkg/statestore/mock"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/topology/full"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// TestSencChunk tests that a chunk that is uploaded to localstore is sent to the appropriate closest peer.
+func TestSendChunk(t *testing.T) {
+	logger := logging.New(ioutil.Discard, 0)
+
+	// chunk data to upload
+	chunkAddress := swarm.MustParseHexAddress("7000000000000000000000000000000000000000000000000000000000000000")
+	chunkData := []byte("1234")
+
+	// create a pivot node and a cluster of nodes
+	pivotNode := swarm.MustParseHexAddress("0000000000000000000000000000000000000000000000000000000000000000") // base is 0000
+	connectedPeers := []p2p.Peer{
+		{
+			Address: swarm.MustParseHexAddress("8000000000000000000000000000000000000000000000000000000000000000"), // binary 1000 -> po 0
+		},
+		{
+			Address: swarm.MustParseHexAddress("4000000000000000000000000000000000000000000000000000000000000000"), // binary 0100 -> po 1
+		},
+		{
+			Address: swarm.MustParseHexAddress("6000000000000000000000000000000000000000000000000000000000000000"), // binary 0110 -> po 1
+		},
+	}
+	closestPeer := connectedPeers[2].Address
+
+	// mock a connectivity between the nodes
+	p2ps := p2pmock.New(p2pmock.WithConnectFunc(func(ctx context.Context, addr ma.Multiaddr) (swarm.Address, error) {
+		return pivotNode, nil
+	}), p2pmock.WithPeersFunc(func() []p2p.Peer {
+		return connectedPeers
+	}))
+
+	// Create a full connectivity between the peers
+	discovery := mock.NewDiscovery()
+	statestore := mockstate.NewStateStore()
+	ab := addressbook.New(statestore)
+	fullDriver := full.New(discovery, ab, p2ps, logger, pivotNode)
+
+	storer, err := localstore.New("", pivotNode.Bytes(), nil, logger)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// setup the stream recorder to record stream data
+	recorder := streamtest.New(
+		streamtest.WithMiddlewares(func(f p2p.HandlerFunc) p2p.HandlerFunc {
+			return func(context.Context, p2p.Peer, p2p.Stream) error {
+				// dont call any handlers
+				return nil
+			}
+		}),
+	)
+
+	// instantiate a pushsync instance
+	ps := pushsync.New(pushsync.Options{
+		Streamer:      recorder,
+		Logger:        logger,
+		ClosestPeerer: fullDriver,
+		Storer:        storer,
+	})
+	defer ps.Close()
+	recorder.SetProtocols(ps.Protocol())
```

Comment:
Why is protocol being sent on the same recorder? Sorry, I might misunderstand something cuz I am not familiar with the protocol 🙈

Reply:
recorder in its current form is not suitable for this kind of testing that expects a protocol to initiate the creation of its own streams, that is, there is a mutual dependency between recorder and the protocol instance.

the pr description mentions that a simpler recorder implementation is needed, since recorder in its current form actually needs two protocol instances, but this is not needed for this test vector, since there is no initiating "other" node which creates the stream, only one unit is under test here. if you have any other suggestion i am open to hear it out.

Reply:
I agree, but, since we will have receipts, I think that the two sided recorder can be used fully in these tests and maybe shorten them a bit, compared to explicitly testing handler or client function.

Reply:
That's what I meant. Basically, we created streamtest very opinionated, to test both sides and not just one side. If you want to test one side you could basically just mock read, write and create stream calls, and no need to use streamtest at all. Do you think guys it would make sense to create mocks for new stream, read and write calls, and then use them if we want to test just one side, either handler or sender? It will be also easier when we abstract creating streams in general

Reply:
Sorry, maybe it's not the scope of this PR just a general thought :)